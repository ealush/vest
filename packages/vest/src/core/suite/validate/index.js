import promisify from '../../../utilities/promisify';
import create from '../create';

/**
 * Creates a suite and immediately invokes it.
 * @param {string} suiteName
 * @param {Function} tests
 */
export default function validate(...args) {
  return promisify(create(...args))();
}
