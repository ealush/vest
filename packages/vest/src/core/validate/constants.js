/**
 * @type {String}
 */
export const SUITE_INIT_ERROR = "Suite initialization error.";
