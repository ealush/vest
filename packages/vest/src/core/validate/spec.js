import faker from "faker";
import { noop } from "lodash";
import validate from ".";

describe("Test validate suite wrapper", () => {
  describe("Test arguments", () => {
    let mockThrowError, validate;

    beforeEach(() => {
      mockThrowError = jest.fn();
      jest.resetModules();
      jest.mock("../../lib/throwError/", () => ({
        __esModule: true,
        default: mockThrowError,
      }));
      validate = require(".");
    });

    afterEach(() => {
      jest.resetAllMocks();
    });

    it.each([[1, {}, noop]])(
      "Should throw a typerror error for a non string name.",
      (value) => {
        validate(value, noop);
        expect(mockThrowError).toHaveBeenCalledWith(
          "Suite initialization error. Expected name to be a string.",
          TypeError
        );
      }
    );

    it.each([[1, {}, "noop"]])(
      "Should throw a typerror error for a non function tests callback.",
      (value) => {
        validate(faker.random.word(), value);
        expect(mockThrowError).toHaveBeenCalledWith(
          "Suite initialization error. Expected tests to be a function.",
          TypeError
        );
      }
    );
  });

  it("Calls `tests` argument", (done) => {
    validate("FormName", done);
  });

  describe("Context creation", () => {
    let mockRunWithContext, validate, name;

    beforeEach(() => {
      name = "formName";
      mockRunWithContext = jest.fn();
      jest.resetModules();
      jest.mock("../../lib/runWithContext", () => ({
        __esModule: true,
        default: mockRunWithContext,
      }));
      validate = require(".");
      validate(name, noop);
    });

    afterEach(() => {
      jest.resetAllMocks();
    });

    it("Should call `runWithContext` with tests as the argument", () => {
      expect(mockRunWithContext.mock.calls[0]).toMatchSnapshot();
    });
  });
});
