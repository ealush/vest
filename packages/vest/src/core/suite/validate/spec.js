import faker from 'faker';
import mock from '../../../../../shared/testUtils/mock';
import { OPERATION_MODE_STATELESS } from '../../constants';
import Context from '../Context';

let validate;

describe('module:validate', () => {
  let suiteName,
    tests,
    mockCreateSuite,
    returnedFn,
    mockCleanupCompletedSuite,
    output,
    suiteId;
  beforeEach(() => {
    returnedFn = jest.fn();
    tests = jest.fn();
    mockCleanupCompletedSuite = mock('cleanupCompletedSuite');
    suiteName = faker.lorem.word();
    suiteId = faker.random.number();
    mockCreateSuite = mock('createSuite', returnedFn);
    mock('id', () => suiteId);
    validate = require('.');
    output = validate(suiteName, tests);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('Test arguments', () => {
    let mockValidateSuiteParams, name, tests;

    beforeEach(() => {
      mockValidateSuiteParams = mock('validateSuiteParams');
      validate = require('.');
      name = faker.random.word();
      tests = jest.fn();
    });

    it('Should call `validateSuiteParams` with passed arguments and current function name', () => {
      expect(mockValidateSuiteParams).not.toHaveBeenCalled();
      validate(name, tests);
      expect(mockValidateSuiteParams).toHaveBeenCalledWith(
        'validate',
        name,
        tests
      );
    });
  });

  it('Should set correct context for test run', () => {
    returnedFn = jest.fn(() => {
      expect(Context.use()).toEqual({
        name: suiteName,
        suite_id: suiteId,
        operationMode: OPERATION_MODE_STATELESS,
      });
    });
    mockCreateSuite = mock('createSuite', returnedFn);
    mock('id', () => suiteId);
    validate = require('.');
    validate(suiteName, tests);
    expect(returnedFn).toHaveBeenCalled();
  });

  it('Should call `createSuite` with passed arguments', () => {
    expect(mockCreateSuite).toHaveBeenCalledWith(suiteName, tests);
  });

  it("Should call `createSuite`'s returned function", () => {
    expect(returnedFn).toHaveBeenCalled();
  });

  it('Should return the returned function output', () => {
    expect(output).toBe(returnedFn.mock.results[0][0]);
  });

  it('Should call all functions in order', () => {
    const unsorted = [
      returnedFn.mock.invocationCallOrder,
      mockCreateSuite.mock.invocationCallOrder,
      mockCleanupCompletedSuite.mock.invocationCallOrder,
    ];

    expect(unsorted.sort()).toEqual([
      mockCreateSuite.mock.invocationCallOrder,
      returnedFn.mock.invocationCallOrder,
      mockCleanupCompletedSuite.mock.invocationCallOrder,
    ]);
  });

  describe('When `createSuite` does not return a function', () => {
    it('Should return silently', () => {
      const output = validate('');
      expect(output).toBeUndefined();
    });
  });
});
