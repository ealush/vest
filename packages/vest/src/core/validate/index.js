import { throwError, runWithContext } from "../../lib";
import suiteResult from "../suiteResult";
import { runAsync } from "../test";
import { SUITE_INIT_ERROR } from "./constants";

/**
 * Initializes a validation suite, creates a validation context.
 * @param {String} name     Descriptive name for validation suite.
 * @param {Function} tests  Validation suite body.
 * @returns {Object} Vest output object.
 */
const validate = (name, tests) => {
  if (typeof name !== "string") {
    return throwError(
      SUITE_INIT_ERROR + " Expected name to be a string.",
      TypeError
    );
  }

  if (typeof tests !== "function") {
    return throwError(
      SUITE_INIT_ERROR + " Expected tests to be a function.",
      TypeError
    );
  }

  const result = suiteResult(name);

  runWithContext({ result }, () => {
    tests();
    [...result.pending].forEach(runAsync);
  });

  return result.output;
};

export default validate;
