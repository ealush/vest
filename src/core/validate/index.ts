import { throwError } from '../../lib';
import Context from '../Context';
import { runAsync } from '../test';
import suiteResult, { VestOutput } from '../suiteResult';
import { SUITE_INIT_ERROR } from './constants';

/**
 * Initializes a validation suite, creates a validation context.
 */
const validate = (name: string, tests: Function): VestOutput => {
    if (typeof name !== 'string') {
        //@ts-ignore
        return throwError(SUITE_INIT_ERROR + ' Expected name to be a string.', TypeError);
    }

    if (typeof tests !== 'function') {
        //@ts-ignore
        return throwError(SUITE_INIT_ERROR + ' Expected tests to be a function.', TypeError);
    }

    const result = suiteResult(name);

    new Context({ result });

    tests();

    Context.clear();

    [...result.pending].forEach(runAsync);

    return result.output;
};

export default validate;
