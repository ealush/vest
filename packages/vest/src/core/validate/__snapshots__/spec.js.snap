// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test validate suite wrapper Context creation Should call \`runWithContext\` with tests as the argument 1`] = `
Array [
  Object {
    "result": Object {
      "addToSkipped": [Function],
      "markAsDone": [Function],
      "markFailure": [Function],
      "markTestRun": [Function],
      "output": Object {
        "errorCount": 0,
        "name": "formName",
        "skipped": Array [],
        "testCount": 0,
        "tested": Array [],
        "tests": Object {},
        "warnCount": 0,
      },
      "pending": Array [],
      "setPending": [Function],
    },
  },
  [Function],
]
`;
